keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1076,265 @@ the remaining requests in that queue start getting faster service.  In
+
+As a result, kube-apiserver (and etcd) may be able to easily keep with N
+simple in-flight requests (e.g. create or get a single Pod), but will explode
+trying to process N requests listing all the pods in the system at the same
+time.
+
+#### Width of the request
+
+In order to address this problem, we are introducing the concept of `width`
+of the request.  Instead of saying that every request is consuming a single
+unit of concurrency, we allow for a request to consume `<width>` units of
+concurrency while being processed.
+
+While in theory the `width` can be an arbitrary non-integer number, for
+practical reasons, we will assume it actually is an integer.  Given that
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we are going
+to start with something simple to first get operational experience with it.
+
+While the actual cost of processing a LIST request is roughly proportional
+to the number processed objects, it shouldn't be reflected only in the
+`width` - it will also be affecting the latency of processing the request.
+In practise, processing a single request isn't really parallelized. So
+if we scale the overall concurrency limit with number of available cores,
+the request should never be able to effectively eat more than N units.
+
+Given that we should try to stay on the safe side and rather overprovision
+the number of units needed by the request, just hardcoding them for some
+constant N for all LIST requests seem like a reasonable starting point.
+
+Given that the estimation is well separated piece of logic, we can decide
+to replace with much more sophisticated logic later (e.g. whether it is
+served from etcd or from cache, whether it is namespaced or not, etc.).
+
+One more important aspect to resolve is what happens if a given priority
+level doesn't have enough concurrency units assigned to it. To be on the
+safe side we should probably implement borrowing across priority levels
+but until this is done, we will simply cap the `width` at the concurrency
+units assigned to a priority level.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  While the queueing mechanism doesn't require any adjustments,
+now the request in order to start processing has to accumulate `<width>`
+units of concurrency.
+
+The important requirement to recast now is fairness. As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
```

Comment:
comment for myself: 

currently it's basically "latency of requests processed".
So with width, it should basically be  multiple by width (it's basically implementation of the negative budget idea mentioned below)

I will update it to better describe it.